<!DOCTYPE HTML>
<html>

<body>

<h3>Citation:</h3>
<p>A pre-print of our paper is available on <a href = "https://www.biorxiv.org/content/early/2018/03/28/291187" target = "_blank">bioxRiv</a>.</p>
<p>If you use MENTHU in your research, please cite:</p>
<p>Toward Precision Molecular Surgery: Robust, Selective Induction of Microhomology-mediated End Joining in vivo. Hirotaka Ata, Thomas L Ekstrom, Gabriel Martinez-Galvez, Carla M Mann, Alexey V Dvornikov, Kyle J Schaefbauer, Alvin C Ma, Drena Dobbs, Karl J Clark, Stephen C Ekker. bioRxiv 291187; doi: https://doi.org/10.1101/291187</p>
<br>
<br>
<h3>Abstract: </h3>
<p>One key problem in precision genome editing is the resultant unpredictable plurality of sequence outcomes at the site of targeted DNA double-strand breaks (DSBs). This is due to the typical activation of the versatile Non-homologous End Joining (NHEJ) pathway. Such unpredictability limits the utility of somatic gene editing for applications including gene therapy and functional genomics. For germline editing work, the accurate reproduction of identical alleles using NHEJ is a labor intensive process.</p> 
<p>In this study, we propose inducing Microhomology-mediated End Joining (MMEJ) as a viable solution for improving somatic sequence homogeneity in vivo, capable of generating a single predictable allele at high rates (56% ~ 86% of the entire mutant allele pool). Using a combined dataset from zebrafish (Danio rerio) in vivo and human HeLa cell in vitro as a training dataset, we identified specific contextual sequence determinants surrounding genomic DSBs for robust MMEJ pathway activation. We then applied our observation and prospectively designed MMEJ-inducing sgRNAs against a variety of proof-of-principle genes and demonstrated a high level of mutant allele homogeneity at these loci. F0 mutant zebrafish embryos and larvae generated with these gRNAs faithfully recapitulated previously reported, recessive loss-of-function phenotypes. We also provide a novel algorithm MENTHU (http://genesculpt.org/menthu/) for improved prediction of candidate MMEJ loci, suitable for both targeted and genome-wide applications.</p> 
<p>We believe that this MMEJ-centric approach will have a broad impact on genome engineering and its applications. For example, whereas somatic mosaicism hinders efficient recreation of a knockout mutant allele at base pair resolution via the standard NHEJ-based approach, we demonstrate that F0 founders transmitted the identical MMEJ allele of interest at high rates. Most importantly, the ability to directly dictate the reading frame of an endogenous target will have important implications for gene therapy applications in human genetic diseases.</p>
</body>
</html>
