<!DOCTYPE HTML>
<html>
<body>

<h3 style = "display:inline-block">Instructions  </h3><a href="#instructions" data-toggle = "collapse"> expand/collapse </a>
<div id = "instructions" class = "collapse in">
<p>MENTHU will scan a gene (or specified exons within a gene)</p>
<p>To use MENTHU, choose the "Calculate MENTHU Score" tab.</p>
<p>Step 1) Choose the nuclease(s) you intend to use for generating your knockout. MENTHU will search your submitted sequence (see Step 3) for the PAM sequence(s) of your selected Cas9 species. You can search for all PAM types by clicking 'Select All PAMs', or uncheck all PAM types by clicking 'De-select ALL PAMs'. Note that you MUST choose at least one Cas9 species OR use TALEN targets (see Step 1a).</p>
<p>Step 1a) If you wish to use TALEN targets, choose "Yes" in the radio button options under 1a. You will then be prompted to specify TALEN design parameters. You can choose to search for TALEN arms ranging from 15 to 18 nucleotides in size. If you want to search only for TALEN targets with, for example, 16 nucleotide arms, set the "Min" parameter for TALEN arm length to 16, and the "Max" parameter to 16. You can also choose the length of the spacer region between the TALEN arms; the spacer region must be between 14 and 16 nucleotides in length; if you want to set a single length for the spacer region, set the "Min" and "Max" parameters to the same value.</p>
<p>Step 2) Choose your sequence input type. If using a GenBank gene id, select the "GenBank Gene ID" radio button and then paste the ID into the box. If copy/pasting the gene sequence, select the "Copy/paste gene sequence" radio button and then paste your sequence into the box. If copy/pasting, you will then be asked if your pasted sequence has multiple exons, and if you want to target a specific exon within the sequence. If you choose "No", MENTHU will search for target sites within the entirety of your pasted sequence. If you choose "Yes", MENTHU will display table in which you can enter exon information for your sequence.</p>
<p>The exon table has two columns - one to specify the location of the first nucleotide in the exon (exonStart) and one to specify the location of the last nucleotide in the exon (exonEnd). For example, if you paste a gene sequence that has 300 nucleotides in it, and the first exon runs from the 34th to the 121st nucleotides in the sequence, you would click on the box in the first row under "exonStart" and type "34", then click on the box in the first row under "exonEnd" and type "121". Note that you cannot have negative values in the table, and your "exonEnd" must be smaller than the length of your sequence (your exon can't end after the sequence does.) If you are only interested in one exon, you can enter that information in the first row and leave the remaining rows blank. If you are interested in multiple exons, you can enter that information into the provided rows. If you have more than five exons of interests, you can right-click or CMD+click on the fifth row and choose "Insert row below" to add additional rows.</p>
<p>Step 2a) This step will be displayed if you choose to submit a GenBank gene ID by selecting the "GenBank Gene ID" radio button. You will then be prompted as to which exons in the corresponding GenBank gene should be targeted.</p>
<p>Step 3) Choose a score threshold for reporting results. We highly recommend that you choose to <b>DISPLAY</b> sites with a competition score greater than or equal to 40, but only <b>USE</b> sites with a score greater than or equal to 50. Sites scoring 40-50 may be used if no sites with a score of >50 are present, as many sequences do not have any sites with a score greater than 50. Therefore, if you choose 40 as the threshold cutoff, you will display all targets worth consideration.</p>
</div>
<br>
<h3 style = "display:inline-block">Frequently Asked Questions (FAQs)  </h3><a href="#FAQs" data-toggle = "collapse"> expand/collapse </a>
<div id = "FAQs" class = "collapse in">
<p><b>I used an ENSEMBL/Rfam/Pfam/UNIPROT/PDB/NDB/EMBL/BioCyc/PROSITE ID, and didn't get results. Why?</b></p>
<p>MENTHU currently only supports GenBank/RefSeq accessions, and will only perform an analysis on DNA sequences. Check your input to make sure you're using a GenBank or RefSeq ID, and that the ID corresponds to a DNA entry.</p>
<br>
<p><b>Why is it taking so long for me to get results?</b></p>
<p>In order to generate scores, MENTHU has to perform a series of calculations. The amount of time needed to perform those calculations increases with: </p>
<li>sequence length</li>
<li>number and length of exons</li>
<li>number of target types (e.g., more PAMs = more time)</li>
<li>lower minimum score threshold</li>
<br>
<p>You can decrease the length of time needed to get results by :</p>
<li>limiting your search to the beginning of a sequence</li>
<li>choosing one or two exons in which to search for targets</li>
<li>choosing only one or two nuclease target types</li>
<li>setting a more stringent score threshold (we recommend using the default setting of 40)</li>
<p></p>
<p>We are in the progress of implementing an email submission form for performing longer calculations.</p>
<br>
<p><b>I specified targets in X exon, but the targets returned to me are from a different exon. What gives?</b></p>
<p>This problem can occur if:</p>
<li>Your submitted GenBank/RefSeq accession ID contains multiple genes (which can occur if you submit a genomic section)</li>
<li>The submitted accession is missing one or more exons</li>
<li>The submitted accession has multiple version of the same exon</li>
<p></p>
<p>We are working on a solution to this.</p>
<br>
</div>
</body>
</html>