<!DOCTYPE HTML>
<html>
<body>

<h3 style = "display:inline-block">Instructions  </h3><a href="#instructions" data-toggle = "collapse"> expand/collapse </a>
<div id = "instructions" class = "collapse in">
<p>MENTHU will scan a gene (or specified exons within a gene)</p>
<p>To use MENTHU, choose the "Calculate MENTHU Score" tab.</p>
<!--<p>Step 0) Choose the scoring scheme you would like to use. We recommend Version 2.0, which is the up-to-date methodology. We will continue to support Version 1.0 for the time being, but this may be deprecated in future releases. Legacy versions of MENTHU will always be available via <a href="www.github.com/Dobbs-Lab/menthu">GitHub</a>.</p>-->
<p>Step 1) Choose the nuclease(s) you intend to use for generating your knockout. MENTHU will search your submitted sequence (see Step 3) for the PAM sequence(s) of your selected Cas9 species. You can search for all PAM types by clicking 'Select All PAMs', or uncheck all PAM types by clicking 'De-select ALL PAMs'. Note that you MUST choose at least one Cas9 species OR use TALEN targets (see Step 1a).</p>
<!--<p>Step 1) Choose the nuclease(s) you intend to use for generating your knockout. MENTHU will search your submitted sequence (see Step 3) for the PAM sequence(s) of your selected Cas9 species. You can search for all PAM types by clicking 'Select All PAMs', or uncheck all PAM types by clicking 'De-select ALL PAMs'. Note that you MUST choose at least one Cas9 species (see Step 1a).</p>-->
<p>Step 1a) If you are interested in using a nuclease not listed under (1), you can design a 'custom' PAM sequence. Select the 'Yes' radio button under '1a. Do you want to create a 'custom' PAM sequence?'. Then input your PAM sequence in the first box, and the DSB location in relation to the PAM sequence in the second box. For example, SpCas9's PAM sequence would be "NGG", and its DSB location would be "-3", since SpCas9 cuts three nucleotides upstream of the beginning of its PAM. You can use <a href = "https://www.bioinformatics.org/sms/iupac.html">IUPAC standard nucleotide abbreviations</a> to indicate ambiguity in a PAM sequence. You can input multiple sequences and multiple DSB sites by separating the entries with a comma or space in their appropriate boxes.</p>
<p>Step 1b) If you wish to use TALEN targets, choose "Yes" in the radio button options under 1a. You will then be prompted to specify TALEN design parameters. You can choose to search for TALEN arms ranging from 15 to 18 nucleotides in size. If you want to search only for TALEN targets with, for example, 16 nucleotide arms, set the "Min" parameter for TALEN arm length to 16, and the "Max" parameter to 16. You can also choose the length of the spacer region between the TALEN arms. Due to the ambiguity of the DSB site in a spacer region with an odd number of nucleotides, we highly recommend (and only support) using spacer regions of 14 or 16 nucleotides.</p>
<p>Step 2) Choose your sequence input type. If using a GenBank gene id, select the "GenBank Gene ID" radio button and then paste the ID into the box (Please see the FAQs section for a complete listing of supported and unsupported accession types.). If copy/pasting the gene sequence, select the "Copy/paste gene sequence" radio button and then paste your sequence into the box. If copy/pasting, you will then be asked if your pasted sequence has multiple exons, and if you want to target a specific exon within the sequence. If you choose "No", MENTHU will search for target sites within the entirety of your pasted sequence. If you choose "Yes", MENTHU will display table in which you can enter exon information for your sequence.</p>
<p>The exon table has two columns - one to specify the location of the first nucleotide in the exon (exonStart) and one to specify the location of the last nucleotide in the exon (exonEnd). For example, if you paste a gene sequence that has 300 nucleotides in it, and the first exon runs from the 34th to the 121st nucleotides in the sequence, you would click on the box in the first row under "exonStart" and type "34", then click on the box in the first row under "exonEnd" and type "121". Note that you cannot have negative values in the table, and your "exonEnd" must be smaller than the length of your sequence (your exon can't end after the sequence does.) If you are only interested in one exon, you can enter that information in the first row and leave the remaining rows blank. If you are interested in multiple exons, you can enter that information into the provided rows. If you have more than five exons of interests, you can right-click or CMD+click on the fifth row and choose "Insert row below" to add additional rows.</p>
<p>Step 2a) This step will be displayed if you choose to submit a GenBank gene ID by selecting the "GenBank Gene ID" radio button. You will then be prompted as to which exons in the corresponding GenBank gene should be targeted.</p>
<!--<p>Step 3) Choose a score threshold for reporting results. If using the Version 2 scoring system, we recommend you use only use sites with a score greater than or equal to 1.5. If using the legacy Version 1.0 scoring system, we highly recommend that you choose to <b>DISPLAY</b> sites with a competition score greater than or equal to 40 (and so enter '40' in the threshold box), but only <b>USE</b> sites with a score greater than or equal to 50. Sites scoring 40-50 may be used if no sites with a score of >50 are present, as many sequences do not have any sites with a score greater than 50. Therefore, if you choose 40 as the threshold cutoff, you will display all targets worth consideration.</p>-->
<p>At this point you can click the 'Submit' button. Depending on several factors (including whether a GenBank/RefSeq accession was used, the length of sequence to be searched, the number of PAMs selected/specified, etc.), it may take a few seconds to a few minutes to calculate your scores. Please <strong>DO NOT</strong> leave the page until your results are finished calculating.</p>
<p><strong>Interpreting results:</strong></p>
<p>MENTHU will generate a table of results, consisting of 8 columns and a number of rows. Each row corresponds to a single target site in the GenBank/RefSeq or copied/pasted sequence you provided. The first column (Target_Sequence) gives the gRNA sequence required to target a particular site. The PAM sequence is displayed in bold.</p> 
<p>The second column (MENTHU_Score) provides a score indicating the strength of competing microhomologies near the target site. A larger MENTHU score is associated with a stronger predominant deletion outcome (so larger score = better site). We recommend only using sites with a score >1.5 for best results.</p>
<p>The third column (Frame_Shift) gives a 'Yes' or 'No' answer to whether the top predicted deletion pattern will produce an out-of-frame deletion (and thus cause a frameshift knockout.)</p>
<p>The fourth column (Tool_Type) indicates the PAM sequence required to target a site. This is already included in column 1.</p>
<p>The fifth column (strand) indicate if the target is on the 5' (forward) strand or the 3' (complement) strand to the sequence provided.</p>
<p>The sixth column (Exon_ID) gives the exon the target is located in. In cases where a GenBank file does not have properly annotated exon numbers, Exon_ID will simply correspond to the exons as they are listed in the GenBank file.</p>
<p>The seventh column (Cut_Location) gives the position of the nucleotide directly upstream (directly to the left) of the expected double-strand break site produced by the gRNA.</p>
<p>The eighth column (Top_Deletion) gives the top predicted deletion pattern produced by looking at 80bp of context surrounding the DSB site. The dashes in the sequence indicate a deleted nucleotide.</p>
<p>The number of entries displayed can be changed by clicking on the "Show entries" drop-down box in the upper left hand corner of the table. The results can also be sorted by clicking on the column heading on which you would like to sort. The table can also be searched using the search box in the upper right hand corner.</p>
<p>To download your results, click the "Download Results" button. This will generate a comma separated value (CSV) file, which can be opened in any text editing program or in Microsoft Excel.</p>
</div>
<br>
<h3 style = "display:inline-block">Frequently Asked Questions (FAQs)  </h3><a href="#FAQs" data-toggle = "collapse"> expand/collapse </a>
<div id = "FAQs" class = "collapse in">
<p><b>Does MENTHU store data inputs?</b></p>
<p>No. Your data inputs are only available to MENTHU for as long as your MENTHU session is active, and are only used to calculate your results during that time. The file containing your results is only generated when you click the 'Download' button - no results files are actually created or stored on our server. Once you exit the MENTHU browser window, everything you did in the session is completely gone. (The flip side of this is that if you do not save your results, you will have to perform the analysis again.)</p>
<br>
<p><b>What accession types are supported?</b></p>
<p>MENTHU supports the following RefSeq input types:</p>
<li>Nucleotide inputs</li>
<li>mRNA accessions (RefSeq NM or XM prefixes; please note that if you want to use targets at the beginning or end of an exon, mRNAs and other CDS-based GenBank/RefSeq entries may not have enough context to identify these targets. Use a primary DNA sequence instead.</li>
<li>Nucleotide accessions not on the unsupported list (check the GenBank prefixes <a href="https://www.ncbi.nlm.nih.gov/Sequin/acc.html" target = "_blank">here</a>.)</li>
<br>
<p><b>What accession types are not supported?</b></p>
<p>MENTHU DOES NOT support the following accession types: </p>
<li>Non-nucleotide accessions (GenBank accessions with 3 letter prefix followed by 5 numerals, RefSeq accessions with AP, NP, YP, XP, or WP prefixes)</li>
<li>Whole genome sequences (GenBank accession with a 4 letter prefix followed by 6-10 numerals, RefSeq IDs with AC, NC, or NZ prefixes)</li>
<li>(Incomplete) Genomic regions (RefSeq accession with NG prefix)</li>
<li>Contigs and scaffolds (RefSeq accession with NT or NW prefixes)</li>
<li></li>
<p><b>I used an ENSEMBL/Rfam/Pfam/UNIPROT/PDB/NDB/EMBL/BioCyc/PROSITE ID, and didn't get results. Why?</b></p>
<p>MENTHU currently only supports GenBank/RefSeq accessions, and will only perform an analysis on DNA sequences. Check your input to make sure you're using a GenBank or RefSeq ID, and that the ID corresponds to a DNA entry.</p>
<br>
<p><b>Why is it taking so long for me to get results?</b></p>
<p>In order to generate scores, MENTHU has to perform a series of calculations. The amount of time needed to perform those calculations increases with: </p>
<li>sequence length</li>
<li>number and length of exons</li>
<li>number of target types (e.g., more PAMs = more time)</li>
<li>lower minimum score threshold</li>
<br>
<p>You can decrease the length of time needed to get results by :</p>
<li>limiting your search to the beginning of a sequence</li>
<li>choosing one or two exons in which to search for targets</li>
<li>choosing only one or two nuclease target types</li>
<li>setting a more stringent score threshold (we recommend using the default setting of 40)</li>
<p></p>
<p>We are in the progress of implementing an email submission form for performing longer calculations.</p>
<br>
<p><b>I specified targets in X exon, but the targets returned to me are from a different exon. What gives?</b></p>
<p>This problem can occur if:</p>
<li>Your submitted GenBank/RefSeq accession ID contains multiple genes (which can occur if you submit a genomic section)</li>
<li>The submitted accession is missing one or more exons</li>
<li>The submitted accession has multiple version of the same exon</li>
<p></p>
<p>We are working on a solution to this.</p>
<br>
</div>
</body>
</html>
